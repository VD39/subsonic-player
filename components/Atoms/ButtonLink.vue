<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    disabled?: boolean;
    fullWidth?: boolean;
    icon?: Component | Icon;
    iconColor?: string;
    iconPosition?: IconPosition;
    iconSize?: IconSize;
    iconWeight?: IconWeight;
    is?: ButtonLink;
    showText?: boolean;
    type?: string;
  }>(),
  {
    disabled: undefined,
    icon: undefined,
    iconColor: undefined,
    iconPosition: 'left',
    iconSize: 'medium',
    iconWeight: 'regular',
    is: 'button',
    type: 'button',
  },
);

const type = props.is === 'button' ? props.type : undefined;
const isComponent =
  props.is === 'nuxt-link' ? resolveComponent('NuxtLink') : props.is;
</script>

<template>
  <component
    :is="isComponent"
    :class="[
      'centerItems',
      $style.buttonLink,
      {
        [$style.alignRight]: icon && iconPosition === 'right',
        [$style.disabled]: disabled,
        'centerAll fullWidth': fullWidth,
      },
    ]"
    :disabled
    draggable="false"
    :type
    v-bind="$attrs"
  >
    <component
      :is="icon"
      v-if="icon"
      ref="iconComponent"
      aria-hidden="true"
      :class="$style.icon"
      :color="iconColor"
      :size="ICON_SIZE[iconSize]"
      :weight="iconWeight"
    />

    <span
      ref="text"
      :class="[
        'clamp',
        $style.text,
        {
          visuallyHidden: !showText,
        },
      ]"
    >
      <slot />
    </span>
  </component>
</template>

<style module>
.buttonLink {
  position: relative;
  gap: var(--default-space);
  padding: var(--default-space);
  white-space: nowrap;
}

.alignRight {
  flex-direction: row-reverse;
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.icon {
  flex-shrink: 0;

  [opacity='0.2'] {
    opacity: 0.4;
    fill: var(--secondary-font-color);
  }
}

.text {
  display: block;
  margin-bottom: 1px;
}
</style>
