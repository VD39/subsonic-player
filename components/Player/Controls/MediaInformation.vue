<script setup lang="ts">
import IconButton from '@/components/Buttons/IconButton.vue';

const { openModal } = useModal();
const { currentTrack } = useAudioPlayer();

function addPlaylistModal() {
  switch (currentTrack.value.type) {
    case 'track':
      openModal('trackDetailsModal', {
        track: currentTrack.value,
      });
      break;
    case 'podcast':
      openModal('podcastDescriptionModal', {
        description: currentTrack.value.description,
      });
      break;
  }
}
</script>

<template>
  <IconButton
    icon="PhInfo"
    title="Episode information"
    @click="addPlaylistModal"
  >
    Episode information
  </IconButton>
</template>
