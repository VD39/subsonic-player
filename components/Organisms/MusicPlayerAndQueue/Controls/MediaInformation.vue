<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';

const { currentTrack } = useAudioPlayer();
const { openTrackInformationModal } = useMediaInformation();

function openCurrentTrackInformationModal() {
  openTrackInformationModal(currentTrack.value);
}
</script>

<template>
  <ButtonLink
    :icon="ICONS.information"
    title="Media information"
    @click="openCurrentTrackInformationModal"
  >
    Media information
  </ButtonLink>
</template>
