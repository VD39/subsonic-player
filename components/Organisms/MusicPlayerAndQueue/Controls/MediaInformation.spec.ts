import type { VueWrapper } from '@vue/test-utils';

import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import { useAudioPlayerMock } from '@/test/useAudioPlayerMock';
import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import { mount } from '@vue/test-utils';

import MediaInformation from './MediaInformation.vue';

const openTrackInformationModalMock = vi.fn();

mockNuxtImport('useDescription', () => () => ({
  openTrackInformationModal: openTrackInformationModalMock,
}));

const { currentTrackMock } = useAudioPlayerMock();

function factory(props = {}) {
  return mount(MediaInformation, {
    props: {
      ...props,
    },
  });
}

describe('MediaInformation', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when the ButtonLink component is clicked', () => {
    beforeEach(() => {
      wrapper.findComponent(ButtonLink).vm.$emit('click');
    });

    it('calls the openModal function with correct parameters', () => {
      expect(openTrackInformationModalMock).toHaveBeenCalledWith(
        currentTrackMock.value,
      );
    });
  });
});
