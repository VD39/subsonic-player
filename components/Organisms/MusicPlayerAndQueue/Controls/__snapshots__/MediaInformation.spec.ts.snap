// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MediaInformation > matches the snapshot 1`] = `
"<button class="centerItems buttonLink" draggable="false" type="button" title="Media information">
  <phinfo aria-hidden="true" class="icon" size="24" weight="regular"></phinfo><span class="clamp text visuallyHidden"> Media information </span>
</button>"
`;
